File: packages/bot/src/ratelimit.ts
/** Per-user rate limiting. PRD 5.3 #3 — protects API costs and abuse. */

const WINDOW_MS = 60_000;
const MAX_PER_WINDOW = 20;

const buckets = new Map<number, number[]>();

export function allow(userId: number): boolean {
  const now = Date.now();
  const hits = (buckets.get(userId) ?? []).filter((t) => now - t < WINDOW_MS);
  if (hits.length >= MAX_PER_WINDOW) {
    buckets.set(userId, hits);
    return false;
  }
  hits.push(now);
  buckets.set(userId, hits);
  return true;
}

// Periodic cleanup so the map doesn't grow unbounded.
setInterval(() => {
  const now = Date.now();
  for (const [id, hits] of buckets) {
    const live = hits.filter((t) => now - t < WINDOW_MS);
    if (live.length === 0) buckets.delete(id);
    else buckets.set(id, live);
  }
}, WINDOW_MS).unref();

File: crates/oddsports-bot/src/ratelimit.rs
//! Per-user rate limiting. PRD 5.3 #3 — protects API costs and abuse.

use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

const WINDOW: Duration = Duration::from_secs(60);
const MAX_PER_WINDOW: usize = 20;

static BUCKETS: Mutex<Option<HashMap<i64, Vec<Instant>>>> = Mutex::new(None);

pub fn allow(user_id: i64) -> bool {
    let mut guard = BUCKETS.lock().unwrap();
    let buckets = guard.get_or_insert_with(HashMap::new);
    let now = Instant::now();

    let hits = buckets.entry(user_id).or_default();
    hits.retain(|t| now.duration_since(*t) < WINDOW);
    if hits.len() >= MAX_PER_WINDOW {
        return false;
    }
    hits.push(now);

    // Opportunistic cleanup so the map doesn't grow unbounded.
    if buckets.len() > 10_000 {
        buckets.retain(|_, v| v.iter().any(|t| now.duration_since(*t) < WINDOW));
    }
    true
}
